Adds support for Keyboard LED (Caps Lock, Scroll Lock, Num Lock)

PJ789 opened the pull request:
Based on code here, adapted very slightly to improve legibility;-
https://forum.arduino.cc/t/leonardo-keyboard-leds-emulation/169582/23

PJ789 committed:
Update HID.cpp

PJ789 committed:
Update HID.cpp

PJ789 committed:
Update HID.h

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino/ArduinoCore-avr?pullRequest=446) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino/ArduinoCore-avr?pullRequest=446) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino/ArduinoCore-avr?pullRequest=446) it.</sub>

PJ789 commented:
@facchinm please would you consider for merge? (This change is a dependency for a related pull request I'm about to add to the keyboard library, adding support for caps lock, scroll lock, num lock leds).

PJ789 commented:
@facchinm please see also

[Keyboard pull request 61](https://github.com/arduino-libraries/Keyboard/pull/61)

... which has dependencies on this code to add LED status handling to the Keyboard API.

thanks
Pete

PJ789 committed:
Added keybboard led status report callback

Callback function avoids the need to poll for status changes.

PJ789 committed:
Added keyboard led status report callback

Avoids the need to continuously poll for changes to caps lock scroll lock num lock

PJ789 commented:
Also added a callback feature on LED status reports from the host, so to avoid continuously polling the API to discover if an LED has changed state.

Requires programmer to set the callback function as follows;

`
void setup() {
  HID().setKeyboardLedsStatusReportCallback( keyboardLedsStatusReportCallback );
}

void loop() {
}

void keyboardLedsStatusReportCallback()
{
  // code to set LEDs, respond to caps lock status change, etc etc
}

`

per1234 commented:
Thanks @PJ789!

Don't worry about these CI failures:

```
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git fetch --depth=1 --no-tags --prune --recurse-submodules origin master
  stderr: 'fatal: unable to access 'https://github.com/arduino/ArduinoCore-avr/': gnutls_handshake() failed: Error in the pull function.'
```

That error is unrelated to the changes you propose here in your pull request.

I have been seeing transient GitHub Actions workflow run failures all day across many repositories and I suspect this is another. One of the repository maintainers can re-run the failing workflow once GitHub gets back to a working state.

PJ789 commented:
@per1234 that's good to hear. I'm using the code on a current HID keyboard project, where it compiles without issue (Leonardo/SS Micro board), and works well. I'd be delighted to see the capability shared with other developers.

facchinm commented on libraries/HID/src/HID.cpp:
```diff
@@ -133,25 +133,48 @@ bool HID_::setup(USBSetup& setup)
 {
 	epType[0] = EP_TYPE_INTERRUPT_IN;
 	PluggableUSB().plug(this);
+  setKeyboardLedsStatusReportCallback( 0 );
```
I'd replace `0` with `NULL` to make the purpose clearer. Also, please use the same indentation as the rest of the file.

PJ789 replied:
Agreed

facchinm commented on libraries/HID/src/HID.cpp:
```diff
@@ -133,25 +133,48 @@ bool HID_::setup(USBSetup& setup)
 		}
 		if (request == HID_SET_REPORT)
```
Is keyboard LEDs the only time this kind of request is being actively used? I'd prefer to add some more generic API as in PluggableUSB framework

facchinm commented:
The best thing to do, IMO, would be to create a new PluggableHID API where a derived module can register its function (or it's ignored otherwise). Any chance you can try this approach?

PJ789 committed:
Update HID.cpp
